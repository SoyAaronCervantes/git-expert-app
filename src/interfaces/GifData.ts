export interface GifsData {
  id: string;
  title: string;
  url: string;
}
