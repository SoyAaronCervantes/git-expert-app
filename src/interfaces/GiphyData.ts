export interface GiphyData {
  analytics: GiphyDataAnalytics;
  analytics_response_payload: string;
  bitly_gif_url: string;
  bitly_url: string;
  content_url: string;
  embed_url: string;
  id: string;
  images: GiphyImages;
  import_datetime: string;
  is_sticker: number;
  rating: string;
  slug: string;
  source: string;
  source_post_url: string;
  source_tld: string;
  title: string;
  trending_datetime: string;
  type: string;
  url: string;
}


interface GiphyDataAnalytics {
  onclick: GiphyDataAnalyticsEvents;
  onload: GiphyDataAnalyticsEvents;
  onsent: GiphyDataAnalyticsEvents;
}

interface GiphyDataAnalyticsEvents {
  url: string;
}

interface GiphyImages {
  "480w_still": GiphyImagesData;
  downsized: GiphyImagesData;
  downsized_large: GiphyImagesData;
  downsized_medium: GiphyImagesData;
  downsized_small: GiphyImagesData;
  downsized_still: GiphyImagesData;
  fixed_height: GiphyImagesData;
  fixed_height_downsampled: GiphyImagesData;
  fixed_height_small: GiphyImagesData;
  fixed_height_small_still: GiphyImagesData;
  fixed_height_still: GiphyImagesData;
  fixed_width: GiphyImagesData;
  fixed_width_downsampled: GiphyImagesData;
  fixed_width_small: GiphyImagesData;
  fixed_width_small_still: GiphyImagesData;
  fixed_width_still: GiphyImagesData;
  looping: GiphyImagesData;
  original: GiphyImagesData;
  original_mp4: GiphyImagesData;
  original_still: GiphyImagesData;
  preview: GiphyImagesData;
  preview_gif: GiphyImagesData;
  preview_webp: GiphyImagesData;
};

interface GiphyImagesData {
  width: string;
  size: string;
  url: string;
  height: string;
}
