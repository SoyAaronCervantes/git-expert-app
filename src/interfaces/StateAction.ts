import {Dispatch, SetStateAction} from "react";

export interface StateAction<T> {
  [ value: string ]: Dispatch<SetStateAction<T>>
}