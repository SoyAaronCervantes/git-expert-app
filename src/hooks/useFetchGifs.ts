import {useEffect, useState} from "react";
import {getGifs} from "../services/api/ServiceApiGifs";
import {GifsData} from "../interfaces/GifData";

interface FetchState {
  data: GifsData[];
  loading: boolean
}

export const useFetchGifs = ( category: string ) => {
  
  const [state, setState] = useState<FetchState>({
    data: [],
    loading: true
  });
  
  useEffect( () => {
    getGifs( category ).then( images => {
      setState({
        data: [ ...images ],
        loading: false
      })
    });
  }, [ category ])
  
  return state;
}