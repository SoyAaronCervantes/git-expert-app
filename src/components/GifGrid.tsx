import React from 'react';
import PropTypes from 'prop-types';
import {useFetchGifs} from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";

const GifGridComponent = ({ category }: { category: string }) => {
  
  const { data:images, loading } = useFetchGifs( category );

  return (
    <>
      <h3>{ category }</h3>
      { loading && <p>Loading...</p> }
      <div className={ "grid--cards" }>
        {
          images.map( image => (
            <GifGridItem
              key={ image.id }
              item={ image } />
          ))
        }
      </div>
    </>
  );
};

GifGridComponent.propTypes = {
  category: PropTypes.string.isRequired
};

export default GifGridComponent;
