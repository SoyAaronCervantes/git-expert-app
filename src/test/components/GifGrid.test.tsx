import React from 'react';
import { shallow } from 'enzyme';
import { useFetchGifs } from '../../hooks/useFetchGifs';
import { GifsData } from '../../interfaces/GifData';

import GifGrid from '../../components/GifGrid';

jest.mock( '../../hooks/useFetchGifs' );

describe( 'Test <GifGrid/>', () => {
  const category = 'Goku';
  const mockUseFetchGifs = useFetchGifs as jest.MockedFunction<typeof useFetchGifs>;

  test( 'should display correctly', () => {

    mockUseFetchGifs.mockImplementation( category => {
      return { data: [], loading: true }
    } );

    const wrapper = shallow( <GifGrid category={ category }/> );
    expect( wrapper ).toMatchSnapshot();
  } );

  test( 'should display items from useFetchGifs', () => {

    const data: GifsData[] = [
      {
        id: '0',
        url: 'https://lorem.com/img',
        title: 'mockTitle'
      },
      {
        id: '1',
        url: 'https://lorem.com/img',
        title: 'mockTitle__1'
      }
    ];

    mockUseFetchGifs.mockImplementation( category => {
      return { data, loading: false };
    } );

    const wrapper = shallow( <GifGrid category={ category }/> );

    expect( wrapper.find( 'p' ).exists() ).toBe( false );
    expect( wrapper.find( 'GifGridItem' ).length ).toBe( data.length );
  } );

} );