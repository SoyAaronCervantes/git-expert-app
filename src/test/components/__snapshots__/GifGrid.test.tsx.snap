// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <GifGrid/> should display correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInLeft"
  >
    Goku
  </h3>
  <p
    className="animate__animated animate__fadeInLeft"
  >
    Loading...
  </p>
  <div
    className="grid--cards"
  />
</Fragment>
`;
